In [216]:
import pandas as pd
import country_converter as cc

Although the intersection of our datasets goes back to 1971, we will start in 1992 as many countries were previously part of the USSR and thus did not have independent reporting statistics.

While country name standardization can be a problem, thankfully there is a solution. Both datasets contain IOS3 codes, which are universally able to identify geopolitical entities regardless of language or naming convention. Python has a program, country_converter, that will allow us to change these codes back into recognizable names.

# GDP

In [217]:
gdp = pd.read_csv("/Users/matthewcavanaugh/Desktop/GitHub/gdp-trends-from-1992-to-2019/National GDP table.csv")
gdp

Entity Code  Year  GDP (output, multiple price benchmarks)
0       Albania  ALB  1971                               6958496093
1       Albania  ALB  1972                               7246038085
2       Albania  ALB  1973                               7536605468
3       Albania  ALB  1974                               7855124023
4       Albania  ALB  1975                               8164371093
...         ...  ...   ...                                      ...
10103  Zimbabwe  ZWE  2015                              39798644531
10104  Zimbabwe  ZWE  2016                              40963191406
10105  Zimbabwe  ZWE  2017                              44316742187
10106  Zimbabwe  ZWE  2018                              43420898437
10107  Zimbabwe  ZWE  2019                              40826570312

[10108 rows x 4 columns]

In [218]:
gdp.rename(columns={"GDP (output, multiple price benchmarks)": "Adjusted GDP in USD"}, inplace=True)
gdp.rename(columns={"Code": "ISO3"}, inplace=True)
gdp.drop("Entity", axis=1, inplace=True)
gdp

ISO3  Year  Adjusted GDP in USD
0      ALB  1971           6958496093
1      ALB  1972           7246038085
2      ALB  1973           7536605468
3      ALB  1974           7855124023
4      ALB  1975           8164371093
...    ...   ...                  ...
10103  ZWE  2015          39798644531
10104  ZWE  2016          40963191406
10105  ZWE  2017          44316742187
10106  ZWE  2018          43420898437
10107  ZWE  2019          40826570312

[10108 rows x 3 columns]

In [219]:
%%time

import country_converter as cc

# Convert the ISO3 codes to standard short country names (e.g. USA = United States)
gdp['ISO3'] = gdp["ISO3"].apply(lambda x: cc.convert(names=[x], to='name_short'))

# Any unmatched ISO3 codes will be rendered as 'not found'
# Warning messages will be displayed below
# Remove these values from the dataframe
gdp = gdp[gdp["ISO3"] != "not found"]

# Change the name of the column to 'Country'
gdp.rename(columns={"ISO3": "Country"}, inplace=True)

CPU times: user 4min 46s, sys: 1.21 s, total: 4min 47s
Wall time: 4min 47s


In [220]:
# Sort the dataframe first by 'Country', then by 'Year'
gdp = gdp.sort_values(by=["Country", "Year"])
gdp

Country  Year  Adjusted GDP in USD
0       Albania  1971           6958496093
1       Albania  1972           7246038085
2       Albania  1973           7536605468
3       Albania  1974           7855124023
4       Albania  1975           8164371093
...         ...   ...                  ...
10103  Zimbabwe  2015          39798644531
10104  Zimbabwe  2016          40963191406
10105  Zimbabwe  2017          44316742187
10106  Zimbabwe  2018          43420898437
10107  Zimbabwe  2019          40826570312

[10108 rows x 3 columns]

In [223]:
#gdp.to_csv("Cleaned GDP.csv", index=False)

# Population

In [226]:
population = pd.read_csv("/Users/matthewcavanaugh/Desktop/GitHub/gdp-trends-from-1992-to-2019/Population estimates.csv")
population

Index                    Country ISO3  Year  Population (thousands)  \
0          1                      World  NaN  1950             2493092.848   
1          2                      World  NaN  1951             2536927.035   
2          3                      World  NaN  1952             2584086.339   
3          4                      World  NaN  1953             2634106.235   
4          5                      World  NaN  1954             2685894.860   
...      ...                        ...  ...   ...                     ...   
21973  21979  Wallis and Futuna Islands  WLF  2019                  11.686   
21974  21980  Wallis and Futuna Islands  WLF  2020                  11.616   
21975  21981  Wallis and Futuna Islands  WLF  2021                  11.568   
21976  21982  Wallis and Futuna Islands  WLF  2022                  11.478   
21977  21983  Wallis and Futuna Islands  WLF  2023                  11.370   

       Male population (thousands)  Female population (thousands)  \
0                      1241431.646                    1251661.202   
1                      1263362.496                    1273564.539   
2                      1287092.948                    1296993.391   
3                      1312354.748                    1321751.488   
4                      1338531.144                    1347363.715   
...                            ...                            ...   
21973                        5.629                          6.057   
21974                        5.587                          6.029   
21975                        5.555                          6.013   
21976                        5.495                          5.984   
21977                        5.420                          5.951   

       Population density (sq km)  Sex ratio (males per 100 females)  \
0                          19.118                             99.183   
1                          19.454                             99.199   
2                          19.816                             99.237   
3                          20.200                             99.289   
4                          20.597                             99.344   
...                           ...                                ...   
21973                      83.471                             92.934   
21974                      82.971                             92.653   
21975                      82.625                             92.391   
21976                      81.986                             91.828   
21977                      81.218                             91.069   

       Median age  Population Change (thousands)  Population growth rate  \
0          22.159                      43337.691                   1.738   
1          22.123                      44330.684                   1.747   
2          22.074                      49987.925                   1.935   
3          22.011                      50051.867                   1.900   
4          21.937                      53525.382                   1.993   
...           ...                            ...                     ...   
21973      35.056                         -0.108                  -0.924   
21974      35.590                         -0.032                  -0.275   
21975      36.085                         -0.065                  -0.562   
21976      36.599                         -0.114                  -0.993   
21977      37.166                         -0.101                  -0.888   

       Births (thousands)  Births by women aged 15 to 19 (thousands)  \
0               91823.936                                  10492.124   
1               92507.128                                  10573.672   
2               97371.459                                  11033.785   
3               97291.452                                  11045.162   
4              100187.854                                  11301.725   
...                   ...      

In [227]:
# 'Index' was mistakenly made a column at some point point, drop it
population.drop("Index", axis=1, inplace=True)

# Ensure "Year" is numeric
population["Year"] = population["Year"].astype("Int64")

# Drop the years outside our desired range (1971 to 2019)
population = population[(population["Year"] >= 1971) & (population["Year"] <= 2019)]

# Drop rows where the ISO3 is blank
population = population.dropna(subset=["ISO3"])

population

Country ISO3  Year  Population (thousands)  \
2611                     Burundi  BDI  1971                3617.289   
2612                     Burundi  BDI  1972                3617.128   
2613                     Burundi  BDI  1973                3611.991   
2614                     Burundi  BDI  1974                3703.590   
2615                     Burundi  BDI  1975                3796.612   
...                          ...  ...   ...                     ...   
21969  Wallis and Futuna Islands  WLF  2015                  12.160   
21970  Wallis and Futuna Islands  WLF  2016                  12.033   
21971  Wallis and Futuna Islands  WLF  2017                  11.918   
21972  Wallis and Futuna Islands  WLF  2018                  11.800   
21973  Wallis and Futuna Islands  WLF  2019                  11.686   

       Male population (thousands)  Female population (thousands)  \
2611                      1765.972                       1851.318   
2612                      1764.218                       1852.909   
2613                      1759.612                       1852.379   
2614                      1805.481                       1898.110   
2615                      1852.488                       1944.123   
...                            ...                            ...   
21969                        5.888                          6.272   
21970                        5.821                          6.212   
21971                        5.760                          6.158   
21972                        5.694                          6.106   
21973                        5.629                          6.057   

       Population density (sq km)  Sex ratio (males per 100 females)  \
2611                      139.395                             95.390   
2612                      139.388                             95.213   
2613                      139.190                             94.992   
2614                      142.720                             95.120   
2615                      146.305                             95.287   
...                           ...                                ...   
21969                      86.857                             93.862   
21970                      85.950                             93.706   
21971                      85.129                             93.537   
21972                      84.289                             93.245   
21973                      83.471                             92.934   

       Median age  Population Change (thousands)  Population growth rate  \
2611       16.235                         87.589                   2.422   
2612       15.973                        -87.913                  -2.431   
2613       15.736                         77.640                   2.150   
2614       15.841                        105.559                   2.850   
2615       15.882                         80.484                   2.120   
...           ...                            ...                     ...   
21969      32.693                         -0.138                  -1.135   
21970      33.284                         -0.116                  -0.964   
21971      33.854                         -0.114                  -0.957   
21972      34.455                         -0.121                  -1.025   
21973      35.056                         -0.108                  -0.924   

       Births (thousands)  Births by women aged 15 to 19 (thousands)  \
2611              173.238                                     18.478   
2612              176.528                                     18.228   
2613              168.491                                     17.359   
2614              174.302                                     17.753   
2615              181.986                                     18.356   
...                   ...                                        ...   
21969               0.154                                    

In [228]:
%%time

import country_converter as cc

# Convert the ISO3 codes to standard short country names (e.g. USA = United States)
population["ISO3"] = population["ISO3"].apply(lambda x: cc.convert(names=[x], to='name_short'))

# Any unmatched ISO3 codes will be rendered as 'not found'
# Warning messages will be displayed below
# Remove these values from the dataframe
population = population[population["ISO3"] != 'not found']

# Drop our old 'Country' column
population.drop("Country", axis=1, inplace=True)

# Change the name of the ISO3 column to 'Country'
population.rename(columns={"ISO3": "Country"}, inplace=True)

CPU times: user 5min 29s, sys: 1.45 s, total: 5min 30s
Wall time: 5min 31s


In [229]:
# Sort the dataframe first by 'Country', then by 'Year'
population = population.sort_values(by=['Country', 'Year'])
population

Country  Year  Population (thousands)  Male population (thousands)  \
8457  Afghanistan  1971               11567.667                     5925.771   
8458  Afghanistan  1972               11853.696                     6067.256   
8459  Afghanistan  1973               12157.999                     6218.247   
8460  Afghanistan  1974               12469.127                     6372.901   
8461  Afghanistan  1975               12773.954                     6524.498   
...           ...   ...                     ...                          ...   
4061     Zimbabwe  2015               14399.013                     6810.139   
4062     Zimbabwe  2016               14600.294                     6909.792   
4063     Zimbabwe  2017               14812.482                     7015.107   
4064     Zimbabwe  2018               15034.452                     7125.471   
4065     Zimbabwe  2019               15271.368                     7243.442   

      Female population (thousands)  Population density (sq km)  \
8457                       5641.896                      17.816   
8458                       5786.440                      18.256   
8459                       5939.752                      18.725   
8460                       6096.225                      19.204   
8461                       6249.456                      19.674   
...                             ...                         ...   
4061                       7588.874                      37.221   
4062                       7690.502                      37.741   
4063                       7797.375                      38.290   
4064                       7908.981                      38.864   
4065                       8027.926                      39.476   

      Sex ratio (males per 100 females)  Median age  \
8457                            105.032      16.841   
8458                            104.853      16.713   
8459                            104.689      16.591   
8460                            104.538      16.474   
8461                            104.401      16.360   
...                                 ...         ...   
4061                             89.738      16.729   
4062                             89.848      16.772   
4063                             89.968      16.878   
4064                             90.093      17.022   
4065                             90.228      17.187   

      Population Change (thousands)  Population growth rate  \
8457                        275.483                   2.382   
8458                        296.574                   2.502   
8459                        312.033                   2.567   
8460                        310.221                   2.488   
8461                        299.434                   2.344   
...                             ...                     ...   
4061                        193.978                   1.347   
4062                        208.584                   1.429   
4063                        215.792                   1.457   
4064                        228.148                   1.518   
4065                        245.683                   1.609   

      Births (thousands)  Births by women aged 15 to 19 (thousands)  \
8457             593.459                                     78.232   
8458             607.433                                     80.313   
8459             623.433                                     82.362   
8460             640.572                                     84.579   
8461             655.576                                     86.484   
...                  ...                                        ...   
4061             488.386                                     80.068   
4062             477.484                                     81.397   
4063             471.228                                     82.627   
4064             470.991                                     84.606   
4065             475.267                     

In [230]:
#population.to_csv("Cleaned population.csv", index=False)

# Combining the datasets

In [1]:
import pandas as pd
gdp = pd.read_csv("Cleaned GDP.csv")
population = pd.read_csv("Cleaned population.csv")

In [2]:
# First, find the common countries across both DataFrames
common_countries = set(gdp['Country']) & set(population['Country'])
                                             
# Filter each DataFrames to keep only the common countries
population_filtered = population[population['Country'].isin(common_countries)]
gdp_filtered = gdp[gdp['Country'].isin(common_countries)]

# Merge all DataFrames on 'Country' and 'Year'
popandgdp = gdp_filtered.merge(population_filtered, on=['Country', 'Year'], how='inner')

# Display our combined DataFrame to ensure these changes were properly applied
popandgdp

Country  Year  Adjusted GDP in USD  Population (thousands)  \
0      Albania  1971           6958496093                2234.992   
1      Albania  1972           7246038085                2289.265   
2      Albania  1973           7536605468                2344.105   
3      Albania  1974           7855124023                2399.401   
4      Albania  1975           8164371093                2454.859   
...        ...   ...                  ...                     ...   
8221  Zimbabwe  2015          39798644531               14399.013   
8222  Zimbabwe  2016          40963191406               14600.294   
8223  Zimbabwe  2017          44316742187               14812.482   
8224  Zimbabwe  2018          43420898437               15034.452   
8225  Zimbabwe  2019          40826570312               15271.368   

      Male population (thousands)  Female population (thousands)  \
0                        1130.886                       1104.106   
1                        1158.899                       1130.367   
2                        1187.153                       1156.951   
3                        1215.590                       1183.812   
4                        1244.047                       1210.812   
...                           ...                            ...   
8221                     6810.139                       7588.874   
8222                     6909.792                       7690.502   
8223                     7015.107                       7797.375   
8224                     7125.471                       7908.981   
8225                     7243.442                       8027.926   

      Population density (sq km)  Sex ratio (males per 100 females)  \
0                         81.575                            102.425   
1                         83.556                            102.524   
2                         85.558                            102.610   
3                         87.576                            102.684   
4                         89.600                            102.745   
...                          ...                                ...   
8221                      37.221                             89.738   
8222                      37.741                             89.848   
8223                      38.290                             89.968   
8224                      38.864                             90.093   
8225                      39.476                             90.228   

      Median age  Population Change (thousands)  Population growth rate  \
0         18.059                         54.042                   2.418   
1         18.185                         54.505                   2.381   
2         18.337                         55.174                   2.354   
3         18.515                         55.418                   2.310   
4         18.724                         55.498                   2.261   
...          ...                            ...                     ...   
8221      16.729                        193.978                   1.347   
8222      16.772                        208.584                   1.429   
8223      16.878                        215.792                   1.457   
8224      17.022                        228.148                   1.518   
8225      17.187                        245.683                   1.609   

      Births (thousands)  Births by women aged 15 to 19 (thousands)  \
0                 73.875                                      3.380   
1                 74.682                                      3.447   
2                 74.894                                      3.568   
3                 75.465                                      3.775   
4                 76.181                                      4.052   
...                  ...                                        ...   
8221             488.386                                     80.068   
8222             477.484               

In [3]:
# Drop rows before 1992 and after 2019
popandgdp = popandgdp[(popandgdp['Year'] >= 1992) & (popandgdp['Year'] <= 2019)]

# Ensure 'popandgdp' has the necessary columns
# Drop years outside our desired range
expected_years = set(range(1992, 2020))  # 2020 is exclusive
grouped = popandgdp.groupby("Country")["Year"].agg(set)

# Identify countries with missing years
incomplete_countries = grouped[grouped.apply(lambda x: x != expected_years)].index

# Remove rows with incomplete Country/Year sets
popandgdp = popandgdp[~popandgdp["Country"].isin(incomplete_countries)]

popandgdp

Country  Year  Adjusted GDP in USD  Population (thousands)  \
21     Albania  1992           9767368164                3282.501   
22     Albania  1993          11222619140                3277.825   
23     Albania  1994          12672190429                3269.410   
24     Albania  1995          13669630859                3258.567   
25     Albania  1996          15067375000                3245.678   
...        ...   ...                  ...                     ...   
8221  Zimbabwe  2015          39798644531               14399.013   
8222  Zimbabwe  2016          40963191406               14600.294   
8223  Zimbabwe  2017          44316742187               14812.482   
8224  Zimbabwe  2018          43420898437               15034.452   
8225  Zimbabwe  2019          40826570312               15271.368   

      Male population (thousands)  Female population (thousands)  \
21                       1641.611                       1640.890   
22                       1635.188                       1642.637   
23                       1626.657                       1642.753   
24                       1616.705                       1641.862   
25                       1605.548                       1640.130   
...                           ...                            ...   
8221                     6810.139                       7588.874   
8222                     6909.792                       7690.502   
8223                     7015.107                       7797.375   
8224                     7125.471                       7908.981   
8225                     7243.442                       8027.926   

      Population density (sq km)  Sex ratio (males per 100 females)  \
21                       119.808                            100.044   
22                       119.637                             99.547   
23                       119.330                             99.020   
24                       118.934                             98.468   
25                       118.464                             97.892   
...                          ...                                ...   
8221                      37.221                             89.738   
8222                      37.741                             89.848   
8223                      38.290                             89.968   
8224                      38.864                             90.093   
8225                      39.476                             90.228   

      Median age  Population Change (thousands)  Population growth rate  \
21        23.656                         -2.468                  -0.075   
22        24.056                         -6.885                  -0.210   
23        24.453                         -9.945                  -0.304   
24        24.830                        -11.740                  -0.360   
25        25.174                        -14.038                  -0.433   
...          ...                            ...                     ...   
8221      16.729                        193.978                   1.347   
8222      16.772                        208.584                   1.429   
8223      16.878                        215.792                   1.457   
8224      17.022                        228.148                   1.518   
8225      17.187                        245.683                   1.609   

      Births (thousands)  Births by women aged 15 to 19 (thousands)  \
21                82.030                                      2.651   
22                79.305                                      2.865   
23                77.976                                      3.120   
24                75.805                                      3.218   
25                71.875                                      2.216   
...                  ...                                        ...   
8221             488.386                                     80.068   
8222             477.484               

Reduce the number of columns. We will only need population and GDP to build our remaining metrics.

In [4]:
popandgdp['Population (thousands)'] = popandgdp['Population (thousands)'] * 1000
popandgdp.rename(columns={'Population (thousands)': 'Population'}, inplace=True)
popandgdp['Population'] = popandgdp['Population'].round().astype(int)
popandgdp = popandgdp[['Country', 'Year', 'Population', 'Adjusted GDP in USD']]
popandgdp

Country  Year  Population  Adjusted GDP in USD
21     Albania  1992     3282501           9767368164
22     Albania  1993     3277825          11222619140
23     Albania  1994     3269410          12672190429
24     Albania  1995     3258567          13669630859
25     Albania  1996     3245678          15067375000
...        ...   ...         ...                  ...
8221  Zimbabwe  2015    14399013          39798644531
8222  Zimbabwe  2016    14600294          40963191406
8223  Zimbabwe  2017    14812482          44316742187
8224  Zimbabwe  2018    15034452          43420898437
8225  Zimbabwe  2019    15271368          40826570312

[4564 rows x 4 columns]

With 4564 rows and a span of 1992 to 2019, there should be 163 unique countries and 28 unique years. Let's check.

In [5]:
popandgdp['Country'].nunique()

163

In [6]:
popandgdp['Year'].nunique()

28

Success!

# GDP analysis

In [7]:
popandgdp['GDP per capita'] = popandgdp['GDP per capita'] = popandgdp['Adjusted GDP in USD'] / popandgdp['Population']
popandgdp

/var/folders/cf/4zrkk0lx1nbdx9tpklpk4jc00000gn/T/ipykernel_56183/927744656.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  popandgdp['GDP per capita'] = popandgdp['GDP per capita'] = popandgdp['Adjusted GDP in USD'] / popandgdp['Population']


Country  Year  Population  Adjusted GDP in USD  GDP per capita
21     Albania  1992     3282501           9767368164     2975.587262
22     Albania  1993     3277825          11222619140     3423.800581
23     Albania  1994     3269410          12672190429     3875.986930
24     Albania  1995     3258567          13669630859     4194.982291
25     Albania  1996     3245678          15067375000     4642.288915
...        ...   ...         ...                  ...             ...
8221  Zimbabwe  2015    14399013          39798644531     2763.984207
8222  Zimbabwe  2016    14600294          40963191406     2805.641544
8223  Zimbabwe  2017    14812482          44316742187     2991.851209
8224  Zimbabwe  2018    15034452          43420898437     2888.093190
8225  Zimbabwe  2019    15271368          40826570312     2673.406227

[4564 rows x 5 columns]

In [8]:
popandgdp.isnull().sum()

Country                0
Year                   0
Population             0
Adjusted GDP in USD    0
GDP per capita         0
dtype: int64

In [9]:
# There is an indexing problem unless I do these steps. Will work on simplifying.
year1992 = popandgdp[popandgdp['Year'] == 1992]
year2019 = popandgdp[popandgdp['Year'] == 2019]
year1992copy = year1992.copy()
year1992copy.set_index('Country', inplace=True)
year2019copy = year2019.copy()
year2019copy.set_index('Country', inplace=True)
year1992copy['GDP per capita'] = year1992copy['Adjusted GDP in USD'] / year1992copy['Population']
year2019copy['GDP per capita'] = year2019copy['Adjusted GDP in USD'] / year2019copy['Population']

In [10]:
gdpcomp = pd.DataFrame()
gdpcomp['GDP 1992'] = year1992copy['Adjusted GDP in USD']
gdpcomp['GDP rank 1992'] = gdpcomp['GDP 1992'].rank(ascending=False, method='min').astype(int)
gdpcomp['GDP per capita 1992'] = year1992copy['GDP per capita']
gdpcomp['GDP per capita rank 1992'] = gdpcomp['GDP per capita 1992'].rank(ascending=False, method='min').astype(int)
gdpcomp['GDP 2019'] = year2019copy['Adjusted GDP in USD']
gdpcomp['GDP rank 2019'] = gdpcomp['GDP 2019'].rank(ascending=False, method='min').astype(int)
gdpcomp['GDP per capita 2019'] = year2019copy['GDP per capita']
gdpcomp['GDP per capita rank 2019'] = gdpcomp['GDP per capita 2019'].rank(ascending=False, method='min').astype(int)
gdpcomp['GDP rank change'] = gdpcomp['GDP rank 1992'] - gdpcomp['GDP rank 2019']
gdpcomp['GDP per capita rank change'] = gdpcomp['GDP per capita rank 1992'] - gdpcomp['GDP per capita rank 2019']
gdpcomp['GDP change raw'] = gdpcomp['GDP 2019'] - gdpcomp['GDP 1992']
gdpcomp['GDP change raw'] = gdpcomp['GDP change raw'].round(3)
gdpcomp['GDP per capita change raw'] = gdpcomp['GDP per capita 2019'] - gdpcomp['GDP per capita 1992']
gdpcomp['GDP per capita change raw'] = gdpcomp['GDP per capita change raw'].round(3)
gdpcomp['GDP growth pct'] = ((gdpcomp['GDP 2019'] - gdpcomp['GDP 1992']) / gdpcomp['GDP 1992']) * 100
gdpcomp['GDP growth pct'] = gdpcomp['GDP growth pct'].round(3)
gdpcomp['GDP per capita growth pct'] = ((gdpcomp['GDP per capita 2019'] - gdpcomp['GDP per capita 1992']) / gdpcomp['GDP per capita 1992']) * 100
gdpcomp['GDP per capita growth pct'] = gdpcomp['GDP per capita growth pct'].round(3)
gdpcomp['GDP growth equitability'] = gdpcomp['GDP per capita growth pct'] - gdpcomp['GDP growth pct']
#gdpcomp.set_index('Country', inplace=True)
gdpcomp

GDP 1992  GDP rank 1992  GDP per capita 1992  \
Country                                                       
Albania      9767368164            112          2975.587262   
Algeria    244668703125             33          9188.203554   
Angola      38985789062             70          3138.014553   
Anguilla      251870574            162         28041.702739   
Argentina  283126843750             26          8403.004047   
...                 ...            ...                  ...   
Uruguay     35659320312             73         11357.549319   
Vietnam    103313757812             52          1513.525458   
Yemen        9439888671            116           629.767000   
Zambia       9762194335            113          1193.906859   
Zimbabwe    60167472656             61          5621.709723   

           GDP per capita rank 1992      GDP 2019  GDP rank 2019  \
Country                                                            
Albania                         117   36103042968            113   
Algeria                          69  507487562500             40   
Angola                          114  227855718750             61   
Anguilla                         25     225680526            162   
Argentina                        72  977420562500             25   
...                             ...           ...            ...   
Uruguay                          55   71122789062             88   
Vietnam                         137  724123375000             31   
Yemen                           161   51828058593             99   
Zambia                          145   56783714843             92   
Zimbabwe                         93   40826570312            105   

           GDP per capita 2019  GDP per capita rank 2019  GDP rank change  \
Country                                                                     
Albania           12514.009646                        92               -1   
Algeria           11721.743485                       100               -7   
Angola             7037.877091                       119                9   
Anguilla          15225.023679                        77                0   
Argentina         21733.272331                        63                1   
...                        ...                       ...              ...   
Uruguay           20935.671567                        65              -15   
Vietnam            7451.839424                       117               21   
Yemen              1476.103111                       155               17   
Zambia             3067.095638                       139               21   
Zimbabwe           2673.406227                       141              -44   

           GDP per capita rank change  GDP change raw  \
Country                                                 
Albania                            25     26335674804   
Algeria                           -31    262818859375   
Angola                             -5    188869929688   
Anguilla                          -52       -26190048   
Argentina                           9    694293718750   
...                               ...             ...   
Uruguay                           -10     35463468750   
Vietnam                            20    620809617188   
Yemen                               6     42388169922   
Zambia                              6     47021520508   
Zimbabwe                          -48    -19340902344   

           GDP per capita change raw  GDP growth pct  \
Country                                                
Albania                     9538.422         269.629   
Algeria                     2533.540         107.418   
Angola                      3899.863         484.458   
Anguilla                  -12816.679         -10.398   
Argentina                  13330.268         245.224   
...                              ...             ...   
Uruguay                     9578.122          99.451   
Vietnam                     5938.314         600.897   
Yemen 

Let's look at which countries have had the strongest GDP growth by percentage.

In [11]:
print(f"GDP growth percentage summary:")
print(f" Number of countries with positive scores: {gdpcomp[gdpcomp['GDP growth pct'] > 0].index.nunique()}")
print(f" Number of countries with negative scores: {gdpcomp[gdpcomp['GDP growth pct'] < 0].index.nunique()}")
print(f" Mean score: {gdpcomp['GDP growth pct'].mean().round(3)} (Closest: {gdpcomp['GDP growth pct'].sub(gdpcomp['GDP growth pct'].mean()).abs().idxmin()})")
print(f" Median score: {gdpcomp['GDP growth pct'].median().round(3)} ({gdpcomp['GDP growth pct'].sub(gdpcomp['GDP growth pct'].median()).abs().idxmin()})")
print(f" Highest score: {gdpcomp['GDP growth pct'].max().round(3)} ({gdpcomp['GDP growth pct'].idxmax()})")
print(f" Lowest score: {gdpcomp['GDP growth pct'].min()} ({gdpcomp['GDP growth pct'].idxmin()})")

GDP growth percentage summary:
 Number of countries with positive scores: 160
 Number of countries with negative scores: 3
 Mean score: 260.468 (Closest: Türkiye)
 Median score: 184.037 (Georgia)
 Highest score: 2831.032 (Equatorial Guinea)
 Lowest score: -76.679 (Montserrat)


In [12]:
gdpcomp['GDP growth pct'].nlargest(10)

Country
Equatorial Guinea    2831.032
Qatar                2065.349
Ethiopia              847.464
Bahrain               841.667
Laos                  745.554
El Salvador           735.598
Mali                  652.104
Jordan                645.798
Myanmar               637.763
Egypt                 608.297
Name: GDP growth pct, dtype: float64

In [13]:
gdpcomp['GDP growth pct'].nsmallest(10)

Country
Montserrat                       -76.679
Zimbabwe                         -32.145
Anguilla                         -10.398
Aruba                              2.871
Barbados                           6.050
Dominica                           7.051
Ukraine                           10.314
Japan                             25.029
Guinea                            28.865
St. Vincent and the Grenadines    33.606
Name: GDP growth pct, dtype: float64

Now let's see what countries have seen the strong per capita GDP growth.

In [14]:
print(f"GDP per capita percentage growth summary:")
print(f" Number of countries with positive scores: {gdpcomp[gdpcomp['GDP per capita growth pct'] > 0].index.nunique()}")
print(f" Number of countries with negative scores: {gdpcomp[gdpcomp['GDP per capita growth pct'] < 0].index.nunique()}")
print(f" Mean score: {gdpcomp['GDP per capita growth pct'].mean().round(3)} (Closest: {gdpcomp['GDP per capita growth pct'].sub(gdpcomp['GDP per capita growth pct'].mean()).abs().idxmin()})")
print(f" Median score: {gdpcomp['GDP per capita growth pct'].median().round(3)} ({gdpcomp['GDP per capita growth pct'].sub(gdpcomp['GDP per capita growth pct'].median()).abs().idxmin()})")
print(f" Highest score: {gdpcomp['GDP per capita growth pct'].max().round(3)} ({gdpcomp['GDP per capita growth pct'].idxmax()})")
print(f" Lowest score: {gdpcomp['GDP per capita growth pct'].min()} ({gdpcomp['GDP per capita growth pct'].idxmin()})")

GDP per capita percentage growth summary:
 Number of countries with positive scores: 150
 Number of countries with negative scores: 13
 Mean score: 127.905 (Closest: Rwanda)
 Median score: 88.575 (Mozambique)
 Highest score: 796.362 (Equatorial Guinea)
 Lowest score: -52.445 (Zimbabwe)


In [15]:
gdpcomp['GDP per capita growth pct'].nlargest(10)

Country
Equatorial Guinea    796.362
El Salvador          650.094
Myanmar              473.921
Laos                 432.691
China                413.089
Ireland              397.556
Vietnam              392.350
Mongolia             354.843
India                338.500
Lebanon              337.114
Name: GDP per capita growth pct, dtype: float64

In [16]:
gdpcomp['GDP per capita growth pct'].nsmallest(10)

Country
Zimbabwe                   -52.445
Montserrat                 -48.699
Anguilla                   -45.706
United Arab Emirates       -43.168
Aruba                      -33.831
Guinea                     -31.640
DR Congo                   -31.422
Burundi                    -23.138
Belize                     -17.574
Central African Republic   -10.478
Name: GDP per capita growth pct, dtype: float64

Let's took at GDP growth equitability, which measures the difference between total and per capita GDP growth. A positive score means per capita growth outpaced total growth, and a negative score the opposite.

In [17]:
print(f"Growth Equitability Score summary:")
print(f" Number of countries with positive scores: {gdpcomp[gdpcomp['GDP growth equitability'] > 0].index.nunique()}")
print(f" Number of countries with negative scores: {gdpcomp[gdpcomp['GDP growth equitability'] < 0].index.nunique()}")
print(f" Mean score: {gdpcomp['GDP growth equitability'].mean().round(3)} (Closest: {gdpcomp['GDP growth equitability'].sub(gdpcomp['GDP growth equitability'].mean()).abs().idxmin()})")
print(f" Median score: {gdpcomp['GDP growth equitability'].median().round(3)} ({gdpcomp['GDP growth equitability'].sub(gdpcomp['GDP growth equitability'].median()).abs().idxmin()})")
print(f" Highest score: {gdpcomp['GDP growth equitability'].max().round(3)} ({gdpcomp['GDP growth equitability'].idxmax()})")
print(f" Lowest score: {gdpcomp['GDP growth equitability'].min()} ({gdpcomp['GDP growth equitability'].idxmin()})")

Growth Equitability Score summary:
 Number of countries with positive scores: 19
 Number of countries with negative scores: 144
 Mean score: -132.563 (Closest: Peru)
 Median score: -84.266 (Suriname)
 Highest score: 112.027 (Georgia)
 Lowest score: -2034.67 (Equatorial Guinea)


In [18]:
gdpcomp['GDP growth equitability'].nlargest(10)

Country
Georgia       112.027
Moldova       106.901
Armenia        78.688
Lithuania      75.284
Latvia         72.507
Romania        56.694
Albania        50.927
Croatia        47.323
Estonia        47.030
Montserrat     27.980
Name: GDP growth equitability, dtype: float64

In [19]:
gdpcomp['GDP growth equitability'].nsmallest(10)

Country
Equatorial Guinea   -2034.670
Qatar               -1801.875
Bahrain              -594.485
Ethiopia             -522.972
Jordan               -457.337
Mali                 -411.060
Angola               -360.180
Oman                 -339.578
Liberia              -335.768
Zambia               -324.774
Name: GDP growth equitability, dtype: float64

Growth has been strong overall. Out of 163 countries, 160 have grown their economies in absolute terms, and 150 have seen the average person get richer. Despite this, only 19 countries have strong per capita than total growth. This creates some discrepancies.

For example, Equatorial Guinea has an extremely low equitability score because its total GDP growth was vastly stronger, despite per capita GDP growth still being the highest in the world at nearly 800%. Thus, -2000 feels too harsh. 

On the other hand, Montserrat has a positive equitability score of about 24. But this is misleading because its total GDP was almost cut in half. This is an especially poor performance considering it is one of just 3 countries to experience absolute contraction.

To better balance all factors, I created a new metric called ['Adjusted growth equitability score'] that balances multiple factors:

A higher ['GDP per capita rank change'] is rewarded, and a lower rank is punished. Using the ranking, rather than raw numbers, allows each country's growth to be fairly measured relative to others. This will prevent countries from receiving high scores if their economy has failed to keep pace with worldwide growth.

A lower ['GDP growth equitability'] is punished, the penalty increasing with a larger ['GDP growth raw']. It wouldn't be fair to measure ['GDP growth raw'] on its own terms considering the difference in population between countries, but it is noteworthy if massive amounts of wealth are created without being shared equitably. This also creates leniency for countries whose economies were smaller to begin with.

I'll also list the rankings for each country.

In [20]:
# Calculate an adjusted growth metric
gdpcomp['Adjusted growth equitability score'] = (
    gdpcomp['GDP per capita rank change']
    - (gdpcomp['GDP growth equitability'] * gdpcomp['GDP change raw'] / gdpcomp['GDP change raw'].max())
).round(3)

# Calculate each country's ranking in the new metric
gdpcomp['Adjusted growth equitability rank'] = gdpcomp['Adjusted growth equitability score'].rank(ascending=False, method='min').astype(int)

Let's see what this new metric looks like.

In [21]:
print(f"Adjusted Growth Equitability Score summary:")
print(f" Number of countries with positive scores: {gdpcomp[gdpcomp['Adjusted growth equitability score'] > 0].index.nunique()}")
print(f" Number of countries with negative scores: {gdpcomp[gdpcomp['Adjusted growth equitability score'] < 0].index.nunique()}")
print(f" Mean score: {gdpcomp['Adjusted growth equitability score'].mean().round(3)} (Closest: {gdpcomp['Adjusted growth equitability score'].sub(gdpcomp['Adjusted growth equitability score'].mean()).abs().idxmin()})")
print(f" Median score: {gdpcomp['Adjusted growth equitability score'].median()} ({gdpcomp['Adjusted growth equitability score'].sub(gdpcomp['Adjusted growth equitability score'].median()).abs().idxmin()})")
print(f" Highest score: {gdpcomp['Adjusted growth equitability score'].max()} ({gdpcomp['Adjusted growth equitability score'].idxmax()})")
print(f" Lowest score: {gdpcomp['Adjusted growth equitability score'].min()} ({gdpcomp['Adjusted growth equitability score'].idxmin()})")

Adjusted Growth Equitability Score summary:
 Number of countries with positive scores: 91
 Number of countries with negative scores: 72
 Mean score: 3.143 (Closest: Trinidad and Tobago)
 Median score: 2.399 (Costa Rica)
 Highest score: 141.498 (China)
 Lowest score: -64.0 (Montserrat)


In [22]:
gdpcomp['Adjusted growth equitability score'].nlargest(15)

Country
China                141.498
India                126.637
Equatorial Guinea     68.880
Qatar                 45.722
El Salvador           39.232
Ireland               36.853
Egypt                 35.588
Myanmar               35.324
Panama                32.531
Bahrain               30.413
Lebanon               30.212
Romania               29.753
Poland                28.996
Singapore             28.827
Ethiopia              28.404
Name: Adjusted growth equitability score, dtype: float64

In [23]:
gdpcomp['Adjusted growth equitability score'].nsmallest(15)

Country
Montserrat                       -64.000
Anguilla                         -52.000
Zimbabwe                         -48.023
Barbados                         -41.000
Dominica                         -41.000
Aruba                            -37.000
Belize                           -35.996
Guinea                           -34.977
Comoros                          -31.996
Algeria                          -29.764
Eswatini                         -28.996
Ukraine                          -26.058
St. Vincent and the Grenadines   -25.000
Gabon                            -23.786
Tunisia                          -20.760
Name: Adjusted growth equitability score, dtype: float64

Intuitively, this feels like a much better measure of economic improvement.

Montserrat has the lowest adjusted score mostly because its per capita GDP rank went from 10th in 1992 to 74th in 2019, the worst dropoff in the world. A mildly "positive" equitability couldn't save it.

Equatorial Guinea, despite extreme inequitability, has the third highest adjusted score at about 68.8. This feels fair because though it has come a long way, it remains relatively poor compared to China and India.

Have a country in mind? Enter its name in the quotes to see all of its statistics!

In [65]:
def lookup_country_stats(df, country):
    if country in df.index:
        print(f"Statistics for {country}:\n" + "-" * (13 + len(country)))
        for col, val in zip(df.columns, df.loc[country].values.tolist()):
            print(f"{col}: {val}")
        print(f"(Ranks are out of 163)")
    else:
        print(f"Country '{country}' not in dataset \n (May have had missing data or does not exist) \n (Check spelling as country name may include special characters)")

In [66]:
# Example usage:
lookup_country_stats(gdpcomp, 'United States')

Statistics for United States:
--------------------------
GDP 1992: 10301173000000.0
GDP rank 1992: 1.0
GDP per capita 1992: 39680.02375168351
GDP per capita rank 1992: 6.0
GDP 2019: 20595844000000.0
GDP rank 2019: 1.0
GDP per capita 2019: 60972.31982845724
GDP per capita rank 2019: 10.0
GDP rank change: 0.0
GDP per capita rank change: -4.0
GDP change raw: 10294671000000.0
GDP per capita change raw: 21292.296
GDP growth pct: 99.937
GDP per capita growth pct: 53.66
GDP growth equitability: -46.277
Adjusted growth equitability score: 24.07
Adjusted growth equitability rank: 24.0
(Ranks are out of 163)
